src/rust/engine/ui/src/console_ui.rs

```diff
@@ -198,40 +109,297 @@ impl ConsoleUI {
+    match renderer.as_str() {
+      "indicatif-spinner" => {
+        // We take exclusive access to stdio by registering a callback that is used to render stderr
+        // while we're holding access. See the method doc.
+        let stderr_dest_bar: Arc<Mutex<Option<WeakProgressBar>>> = Arc::new(Mutex::new(None));
+        // We acquire the lock before taking exclusive access, and don't release it until after
+        // initialization. That way, the exclusive callback can always assume that the destination
+        // is initialized (i.e., can `unwrap` it).
+        let mut stderr_dest_bar_guard = stderr_dest_bar.lock();
+        let (term_read, _, term_stderr_write) = {
+          let stderr_dest_bar = stderr_dest_bar.clone();
+          stdio::get_destination().exclusive_start(Box::new(move |msg: &str| {
+            // Acquire a handle to the destination bar in the UI. If we fail to upgrade, it's because
+            // the UI has shut down: we fail the callback to have the logging module directly log to
+            // stderr at that point.
+            let dest_bar = {
+              let stderr_dest_bar = stderr_dest_bar.lock();
+              // We can safely unwrap here because the Mutex is held until the bar is initialized.
+              stderr_dest_bar.as_ref().unwrap().upgrade().ok_or(())?
+            };
+            dest_bar.println(msg);
+            Ok(())
+          }))?
+        };
+
+        let stderr_use_color = term_stderr_write.use_color;
+        let term = console::Term::read_write_pair_with_style(
+          term_read,
+          term_stderr_write,
+          console::Style::new().force_styling(stderr_use_color),
+        );
+        // NB: We render more frequently than we receive new data in order to minimize aliasing where a
+        // render might barely miss a data refresh.
+        let draw_target = ProgressDrawTarget::term(term, ConsoleUI::render_rate_hz() * 2);
+        let multi_progress = MultiProgress::with_draw_target(draw_target);
+
+        let bars = (0..local_parallelism)
+          .map(|_n| {
+            let style = ProgressStyle::default_bar().template("{spinner} {wide_msg}");
+            multi_progress.add(ProgressBar::new(50).with_style(style))
+          })
+          .collect::<Vec<_>>();
+        *stderr_dest_bar_guard = Some(bars[0].downgrade());
+
+        // Spawn a task to drive the multi progress.
+        let multi_progress_task = executor
+          .spawn_blocking(move || multi_progress.join())
+          .boxed();
+
+        Ok(Instance::Indicatif(IndicatifInstance {
+          tasks_to_display: IndexMap::new(),
+          multi_progress_task,
+          bars,
+        }))
+      }
+      "experimental-prodash" => {
+        // Stderr is propagated across a channel to remove lock interleavings between stdio and the UI.
+        // TODO: Apply a fix similar to #14093 before landing.
```

Comment:
TODO

Reply:
When this is all rebased and otherwise green, I can push a change to do this... or we can do it as a followup change, since this will be experimental for a bit anyway.

Reply:
Let's do a followup? Easier to grok the changes that way.